File: Linked List /Python/2. 142. Linked List Cycle II.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        temp = head
        s = set()
        while temp is not None:
            if temp in s:
                return temp
            else:
                s.add(temp)
                temp = temp.next
        return None

File: Linked List /C++/2. 142. Linked List Cycle II.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        unordered_set <ListNode*> s;
        ListNode *temp = head;
        while(temp != NULL)
        {
            if(s.find(temp) != s.end())
            {
                return temp;
            }
            else
            {
                s.insert(temp);
                temp = temp -> next;
            }
        }
        return NULL;
    }
};
